ttl: do not remove all table TTL settings on RESET (ttl_expire_after)

ecwall opened the pull request:
Fixes #109138

Previously running `RESET (ttl_expire_after)` would remove all table TTL
settings if `ttl_expiration_expression` was set.

This PR fixes the schema changer to only remove all TTL settings on
`RESET (ttl)`.

Release note (bug fix): `RESET (ttl_expire_after)` no longer incorrectly
removes `ttl_expiration_expression`.

ecwall committed:
ttl: do not remove all table TTL settings on RESET (ttl_expire_after)

Fixes #109138

Previously running `RESET (ttl_expire_after)` would remove all table TTL
settings if `ttl_expiration_expression` was set.

This PR fixes the schema changer to only remove all TTL settings on
`RESET (ttl)`.

Release note (bug fix): `RESET (ttl_expire_after)` no longer incorrectly
removes `ttl_expiration_expression`.

ecwall committed:
sql: rename scheduleID to scheduleIDIdx and createStmt to createStmtIdx

scheduleID and createStmt indexes used to populate slice literals for show
results.

Rename these to avoid accidental usage and to prevent shadowing in the sql
package.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110252)
<!-- Reviewable:end -->

ecwall committed:
ttl: refactor row_level_ttl logic tests

Split large subtests into smaller subtests to make debugging easier.

Release note: None

ecwall commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -1241,7 +1241,7 @@ CREATE TABLE public.tbl_set_both_reset_ttl_expire_after (
   CONSTRAINT tbl_set_both_reset_ttl_expire_after_pkey PRIMARY KEY (id ASC),
   FAMILY fam_0_id_expire_at_crdb_internal_expiration (id, expire_at)
-)
+) WITH (ttl = 'on', ttl_expiration_expression = 'expire_at', ttl_job_cron = '@hourly')
```
These changes to the logic test prove the fix works.

ecwall commented:
> One question before looking into the code in detail is it seems like after the change, only `reset(ttl)` will remove the TTL configuration, but what if a table has only `ttl_expire_after` set (so `ttl_expiration_expression` is unset), shouldn't `reset(ttl_expire_after)` also remove the TTL configuration, as we currently do?

It was never intended that `reset(ttl_expire_after)` would reset everything - that was just a bug. To remove all configuration you need to do `reset(ttl)`.

Xiang-Gu commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -1199,3 +1217,75 @@ statement ok
 
 subtest end
+
+subtest set_both_reset_ttl_expire_after
```
nit: can we renmae "set_both_ZZ" to "set_both_xx_and_yy_ZZ" so we know both of what and what are set. Or maybe put this in a comment

ecwall replied:
Done in a commit to make it more readable.

Xiang-Gu commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -1199,3 +1217,75 @@ statement ok
+----
+ACTIVE  @hourly  root
+
+subtest end
```
Overall, I didn't understand why we'd like to put this logictest here in this commit. Shouldn't we add such a logictest in the same commit that we fix the bug so that we can assert that resetting (ttl_expire_after) or resetting (ttl_expiration_expression) do not reset the TTL configuration, but instead only resetting (TTL) will?

ecwall replied:
This subtest was part of another subtest (at least the part that tested the create statement), but it was asserting the wrong thing.

I think it is easier to understand the subtests when they are more focused so I added a commit that splits some subtests up.

Xiang-Gu replied:
I agree with the point of splitting a previously large subtest into smaller ones. What I was saying is the last *newly added* subtest, "subtest set_both_reset_ttl_expire_after" and "set_both_reset_ttl_expiration_expression", shouldn't be in this commit. They should be in the commit where you actually make the fix.

Moreover, subtest "set_both_reset_ttl_expire_after" isn't the expected behavior right? As you said, we'd expect resetting "expire_after" to *not* remove the whole TTL configuration, but the test asserted otherwise (of course, we haven't made any actual code fixes yet so we get the current behavior).

ecwall replied:
Talked about this offline, these were part of subtests above. The subtest that `tbl_set_both_reset_ttl_expire_after` was part of was asserting the wrong thing.

Xiang-Gu commented on pkg/sql/alter_table.go:
I didn't understand the changes in this commit: It seems to me what you've done is mainly on two files:
- `alter_table.go`: you delete the logic to determine whether we should add or drop a TTL and merge it into one -- TTL change. This means we don't set the direction of the TTLMutation.
- `schema_change.go`: as a result of the change in `alter_table.go`, you need to determine what to do when "Done()" on the TTLMutation is called. You will need to plumb in the *after* so you can tell whether you are adding or dropping a mutation.

It seems that those two changes don't have anything to do with the goal of "*only* resetting (TTL) should remove the TTL configuration". I'd prefer we don't make those changes (one reason is we really don't want to use MutationDirection=None). Or, did I miss something here? 

Also, as commented above, can we modify the tests so that they test what you claim to fix (i.e. "resetting ttl_expire_after and ttl_expiration_expression shouldn't reset the whole TTL configuration; only resetting (TTL) should") and move it to this commit?

ecwall replied:
I undid the change to use MutationDirection=None.

chrisseto commented on pkg/sql/catalog/tabledesc/validate.go:
```diff
@@ -562,7 +562,7 @@ func validateMutation(m *descpb.DescriptorMutation) error {
 		}
 	case *descpb.DescriptorMutation_ModifyRowLevelTTL:
-		if m.Direction == descpb.DescriptorMutation_NONE {
+		if m.Direction != descpb.DescriptorMutation_NONE {
```
Could this cause issues if a cluster is upgraded to a version containing this patch while there's a  TTL mutation in flight?

ecwall replied:
I undid this after talking with @Xiang-Gu.

chrisseto commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -595,65 +595,92 @@ ALTER TABLE test.public.tbl_ttl_on_noop WITH (ttl = 'on', ...)
 ----
 ACTIVE  @daily  root
 
+subtest end
```
From my reading of `logic.go` there doesn't appear to be any special handling of `subtest end`. It appears to just create a new subtest called end?

ecwall replied:
Yeah it is just making a subtest called "end", but it is done by convention in order to show where one ends in files that do not specify subtests for all of its subtests. It is only done here for consistency with those files because this file does specify all subtests inside a subtest.

chrisseto commented on pkg/sql/create_table.go:
```diff
@@ -2469,14 +2473,15 @@ func CreateRowLevelTTLScheduledJob(
 ) (*jobs.ScheduledJob, error) {
-	if !tblDesc.HasRowLevelTTL() {
+	if rowLevelTTL == nil {
 		return nil, errors.AssertionFailedf("CreateRowLevelTTLScheduledJob called with no .RowLevelTTL: %#v", tblDesc)
```
nit: Update this error message to match the change to the if statement

ecwall replied:
I undid this change.

ecwall commented on pkg/sql/alter_table.go:
```diff
@@ -1964,10 +1946,22 @@ func handleTTLStorageParamChange(
+	// Adding TTL requires adding the TTL job before adding the TTL fields.
+	// Removing TTL requires removing the TTL job before removing the TTL fields.
+	var direction descpb.DescriptorMutation_Direction
+	switch {
```
Consolidated these cases into a switch to make it clear only one case will happen.

ecwall commented on pkg/sql/schema_changer.go:
```diff
@@ -1595,13 +1595,13 @@ func (sc *SchemaChanger) done(ctx context.Context) error {
 						return err
 					}
 				}
+				scheduledJobs := jobs.ScheduledJobTxn(txn)
```
Only initialize if we are going to use it.

ecwall commented on pkg/sql/schema_changer.go:
```diff
@@ -1642,7 +1642,7 @@ func (sc *SchemaChanger) done(ctx context.Context) error {
 						}
 					}
-					scTable.RowLevelTTL = nil
+					scTable.RowLevelTTL = modify.RowLevelTTL()
```
Do not assume `modify.RowLevelTTL()` is nil even though it will be based on logic to use direction=ADD.

Xiang-Gu reviewed (changes requested):
Good work! I have two last questions:
1. If both `ttl_expire_after` and `ttl_expirataion_expression` are set, and we reset `ttl_expire_after`, the new code path, as I understand it, will be 
 1). Enqueue a DropColumn mutation (so we drop "expire_at" column)
 2). Don't enqueue a TTL mutation but instead directly modify the TTL field on the descriptor
Correct?

2. If we only have `ttl_expire_after` set, and we reset it, it seems to me the same two things above will happen, but maybe later in descriptor validation, we will reject it, right?

Xiang-Gu commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -1251,6 +1251,7 @@ query TTT
 ----
+ACTIVE  @hourly  root
 
 subtest end
```
Can we add one more test case here (under subtest `subtest set_both_reset_ttl_expire_after`) where the `ttl_expirataion_expression` does *not* use the `expire_at` column? We should expect such stmt to succeed, and expect to see that `reset (ttl_expire_after)` drop the `expire_at` column, modify the TTL (but not removing it altogether)?

ecwall replied:
`ttl_expire_after` uses `crdb_internal_expiration` so `reset (ttl_expire_after)` drops `crdb_internal_expiration` not `expire_at`.

ecwall commented:
> Good work! I have two last questions:
> 
> 1. If both `ttl_expire_after` and `ttl_expirataion_expression` are set, and we reset `ttl_expire_after`, the new code path, as I understand it, will be
>    1). Enqueue a DropColumn mutation (so we drop "expire_at" column)
>    2). Don't enqueue a TTL mutation but instead directly modify the TTL field on the descriptor
>    Correct?
> 2. If we only have `ttl_expire_after` set, and we reset it, it seems to me the same two things above will happen, but maybe later in descriptor validation, we will reject it, right?

1-1) It drops `crdb_internal_expiration` not `expire_at`.
1-2) It will not directly modify the descriptor because the mutation will modify the descriptor.
2) You can't reset `ttl_expire_after` if `ttl_expiration_expression` is not set - this did not change and there is a subtest for it.

rafiss reviewed (commented):
lgtm, apart from Xiang's comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110252#-:-NeKG3bY5bDiQMjw2dMQ:btb7d30)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto, @ecwall, and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

ecwall commented:
> > It will not directly modify the descriptor because the mutation will modify the descriptor.
> 
> This is somewhat surprising. Are you sure? In this case, `before != nil && after != nil` so no TTL mutations will be enqueued and the descriptor is modified right away. Or did I miss something?

This code is confusing because it may also be modified in the `before != nil && after != nil` block, but after that it will either add a mutation or change the descriptor immediately. I did not change this behavior.

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110252)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto and @Xiang-Gu)

___
*[`pkg/sql/alter_table.go` line 1958 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110252#gh-1326190089:-NeKIoQW6rWfV_QxnZ7p:b-cv39vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7c4325f21cc411303a5e5f442044eb39462343/pkg/sql/alter_table.go#L1958)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

while you are on this, let's improve around `descriptorChanged` in the following two ways:
1. Rename it to something like `canDirectlyModifyDescriptor` so we know that it's true if we are *not* enqueuing any mutations and can directly modify the descriptor here;
2. Return value from this function will be something like `canDirectlyModifyDescriptor || hasEnqueuedAnyMutation`
</blockquote></details>

I took this name from `descriptorChanged` at the call site. If you want to change this, can you file a separate issue?

The call site documents that `descriptorChanged` should be true only if no mutations are added so it should be changed there also.
```
// Commands can either change the descriptor directly (for  
// alterations that don't require a backfill) or add a mutation to  
// the list.  
descriptorChanged := false
```
___
*[`pkg/sql/alter_table.go` line 1961 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110252#gh-1326184574:-NeKJ2mTA6y0RIRGjFrz:b22uosd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7c4325f21cc411303a5e5f442044eb39462343/pkg/sql/alter_table.go#L1961)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

super nit: "enqueue TTL mutation ..." might be better since the direction might be add or drop.
</blockquote></details>

I took this from the function name I am calling (`AddModifyRowLevelTTLMutation`) the function calling `handleTTLStorageParamChange` also refers to it as "add a mutation" so can you file a separate issue to rename it everwhere?


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110252)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto and @ecwall)

___
*[`pkg/sql/alter_table.go` line 1958 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110252#gh-1326190089:-NeOMOOK1BJjztKhHHZU:b-ltr54q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7c4325f21cc411303a5e5f442044eb39462343/pkg/sql/alter_table.go#L1958)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

I took this name from `descriptorChanged` at the call site. If you want to change this, can you file a separate issue?

The call site documents that `descriptorChanged` should be true only if no mutations are added so it should be changed there also.
```
// Commands can either change the descriptor directly (for  
// alterations that don't require a backfill) or add a mutation to  
// the list.  
descriptorChanged := false
```
</blockquote></details>

I think `descriptorChanged` means is there any updates made to the descriptor. We further make updates into two categories:
1). enqueue a mutation to the descriptor's mutation slice
2). modify other fields directly in the descriptor

In `handleTTLStorageParamChange`, descriptorChanged is not correctly set according to this standard. For example, it's set to false when we enqueue a mutation. That's why I suggested we should rename the variable (within `handleTTLStorageParamChange`) to something like `canDirectlyModifyDescriptor`.  And that's also why the return value of `handleTTLStorageParamChange` should be something like `canDirectlyModifyDescriptor || hasEnqueuedAnyMutation`. Nothing outside `handleTTLStorageParamChange` should change.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110252)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto and @Xiang-Gu)

___
*[`pkg/sql/alter_table.go` line 1958 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/110252#gh-1326190089:-NeOVfx583MiJeCNwCND:bzdxw4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7c4325f21cc411303a5e5f442044eb39462343/pkg/sql/alter_table.go#L1958)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

I think `descriptorChanged` means is there any updates made to the descriptor. We further make updates into two categories:
1). enqueue a mutation to the descriptor's mutation slice
2). modify other fields directly in the descriptor

In `handleTTLStorageParamChange`, descriptorChanged is not correctly set according to this standard. For example, it's set to false when we enqueue a mutation. That's why I suggested we should rename the variable (within `handleTTLStorageParamChange`) to something like `canDirectlyModifyDescriptor`.  And that's also why the return value of `handleTTLStorageParamChange` should be something like `canDirectlyModifyDescriptor || hasEnqueuedAnyMutation`. Nothing outside `handleTTLStorageParamChange` should change.
</blockquote></details>

Please make another ticket to address this with example(s) of where it is being set to true when a mutation is added. Based on the usage of `descriptorChanged`, it only needs to be set if no mutations are added: https://github.com/cockroachdb/cockroach/blob/1755d671f15bfa3a4d111678aca5d022e1a3163a/pkg/sql/alter_table.go#L849


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (approved):
I will take a look at the examples. LGTM! Good work!

Xiang-Gu commented:
@ecwall here you go https://github.com/cockroachdb/cockroach/issues/110727

ecwall commented:
> @ecwall here you go #110727

Great, thanks.
